Bedrock 1.21.60

dries-c opened the pull request:
<!-- Summarize your PR here. Keep it short and simple. -->
<!-- Explain existing problems or why this pull request is necessary -->
This PR updates PM5 to 1.21.60

## Changes
### API changes
<!-- Any additions to the API that should be documented in release notes? -->
<!-- If not, you can delete this section -->
Adds a group argument to adding items to the creative inventory

## Backwards compatibility
<!-- Any possible backwards incompatible changes? How are they solved, or how can they be solved? -->
<!-- If not, you can delete this section -->
See https://github.com/pmmp/BedrockProtocol/pull/294

dries-c committed:
Protocol changes for 1.21.60

dries-c committed:
Codestyle fixes and block state changes

dries-c committed:
Bump block state data version

dries-c committed:
Fix code style

dries-c committed:
Update dependencies

dries-c committed:
Fix door & gate IO

dries-c committed:
Use nbt data from required item list

dries-c committed:
Fix code style

dries-c committed:
Implement creative item groups

dries-c committed:
Fix CI

dries-c committed:
Update dependencies

dktapps commented on src/inventory/CreativeInventory.php:
```diff
@@ -75,15 +101,27 @@ public function clear() : void{
+	 * @return CreativeGroup[]
+	 * @phpstan-return array<int, CreativeGroup>
+	 */
+	public function getItemGroup() : array{
```
```suggestion
	public function getItemGroups() : array{
```

dktapps commented on src/inventory/CreativeInventory.php:
```diff
@@ -75,15 +101,27 @@ public function clear() : void{
+
+	/**
+	 * @return CreativeGroup[]
+	 * @phpstan-return array<int, CreativeGroup>
```
This return value should have a description, it's not obvious how this works from the type alone

dktapps commented on src/inventory/CreativeInventory.php:
```diff
@@ -24,40 +24,65 @@
+			foreach($groups as $groupData){
+				$group = new CreativeGroup(
+					$categoryId,
+					$groupData->group_name,
```
This should use `Translatable`, we don't work on translation keys as plain strings anymore

dktapps commented on src/inventory/CreativeInventory.php:
```diff
@@ -75,15 +101,27 @@ public function clear() : void{
+		return isset($this->items[$index]) ? clone $this->items[$index] : null;
+	}
+
+	public function getGroup(int $index) : ?CreativeGroup{
```
`getItemGroupByIndex` would be a better name. A doc comment with a `@see CreativeInventory::getItemIndex()` would probably be a good idea too.

dktapps commented on src/inventory/CreativeInventory.php:
```diff
@@ -96,9 +134,14 @@ public function getItemIndex(Item $item) : int{
 		$this->onContentChange();
+
+		if($group !== null){ // We need to create a new default group if another group is used.
+			$this->defaultGroup = CreativeGroup::anonymous(CreativeCategory::ITEMS);
```
If several items are added with a defined group, this will create a useless group object for every call.
Better if `defaultGroup` is nullable, and then we only initialize it when it's needed.

dktapps commented on src/inventory/data/CreativeGroup.php:
```diff
@@ -0,0 +1,67 @@
+	private string $name;
+	private ?Item $icon;
+
+	public function __construct(CreativeCategory $categoryId, string $name, ?Item $icon = null){
```
`Translatable|string` should probably be used (although that also depends whether plain non-translated strings are allowed by the client too)

Also not clear why $name is required but $icon is optional

dktapps commented on src/inventory/data/CreativeGroup.php:
```diff
@@ -0,0 +1,67 @@
+		return $this->categoryId;
+	}
+
+	public function setCategoryId(CreativeCategory $categoryId) : void{
```
Don't like this

dktapps commented on src/inventory/data/CreativeGroup.php:
```diff
@@ -0,0 +1,67 @@
+		return $this->name;
+	}
+
+	public function setName(string $name) : void{
```
Don't like this

dktapps commented on src/inventory/data/CreativeGroup.php:
```diff
@@ -0,0 +1,67 @@
+		return $this->icon;
+	}
+
+	public function setIcon(?Item $icon) : void{
```
Don't like this

dktapps commented on src/inventory/data/CreativeGroup.php:
```diff
@@ -0,0 +1,67 @@
+use pocketmine\inventory\CreativeCategory;
+use pocketmine\item\Item;
+
+final class CreativeGroup{
```
This whole class could be simplified down to a few lines using `readonly` properties and promoted constructor parameters

dries-c replied:
readonly classes are PHP 8.2 sadly, so i'll just do the properties for convenience ig

dktapps commented on tools/generate-bedrock-data-from-packets.php:
```diff
@@ -234,31 +253,75 @@ private static function sort(mixed $object) : mixed{
-		}, $packet->getEntries());
-		file_put_contents($this->bedrockDataPath . '/creativeitems.json', json_encode($items, JSON_PRETTY_PRINT) . "\n");
+
+		/** @var array<int, ItemStackData[]> $groupItems */
```
```suggestion
		/**
		 * @var ItemStackData[][] $groupItems
		 * @phpstan-var array<int, list<ItemStackData>> $groupItems
		 */
```

We don't use generics in general @var, they aren't part of the PSR standard

dktapps commented on tools/generate-bedrock-data-from-packets.php:
```diff
@@ -234,31 +253,75 @@ private static function sort(mixed $object) : mixed{
+			CreativeContentPacket::CATEGORY_ITEMS => "items",
+		];
+
+		/** @var array<int, CreativeGroupData[]> $groupCategories */
```
```suggestion
		/**
		 * @var CreativeGroupData[][] $groupCategories
		 * @phpstan-var array<int, list<CreativeGroupData>>
		 */
```
Same as above

Though the doc comment may not be needed at all, try deleting it and see what happens if you do `\PHPStan\dumpType()`

dktapps commented on composer.json:
```diff
@@ -33,10 +33,10 @@
+      "pocketmine/bedrock-data": "dev-bedrock-1.21.60",
       "pocketmine/bedrock-item-upgrade-schema": "~1.14.0+bedrock-1.21.50",
-      "pocketmine/bedrock-protocol": "~35.0.0+bedrock-1.21.50",
+      "pocketmine/bedrock-protocol": "dev-bedrock-1.21.60",
```
These should be normalized to a release version before merge

dries-c commented:
<img width="1280" alt="Screenshot 2025-02-15 at 14 54 48" src="https://github.com/user-attachments/assets/672dd691-5da9-4940-bd9f-df8717f3bdf9" />

We can't do normal translations here since we don't have the player's language. I personally don't think it's a good idea to scrap the entire caching, so how do we best deal with this.

dries-c committed:
Apply Dylan's suggestions

dries-c committed:
Fix PHPStan

dries-c committed:
Move properties into the constructor

dries-c committed:
Resolve translatable's

dries-c committed:
Fix CI

dktapps commented on src/inventory/data/CreativeContentCache.php:
```diff
@@ -0,0 +1,80 @@
+use pocketmine\network\mcpe\protocol\types\inventory\ItemStack;
+use function array_reduce;
+
+final class CreativeContentCache{
```
Not clear why we need a separate class for this.

dries-c replied:
Returning and storing arrays in arrays is less clear

dktapps replied:
I don't see why we need two arrays in the first place.

iAmFrogger commented on src/inventory/data/CreativeContentCache.php:
```diff
@@ -0,0 +1,85 @@
+					}
+
+					if(!$player->getServer()->isLanguageForced()){
+						[$message, ] = $player->getNetworkSession()->prepareClientTranslatableMessage($group->name);
```
Use `$session` instead of `$player->getNetworkSession()` lol

dktapps committed:
Cleanup

dktapps committed:
Rewrite creative inventory handling
we no longer need to specify groups to add items to categories this way. It makes the API make more sense, and also closes off some undefined behaviour potentials.

dktapps committed:
fml

dktapps committed:
Workaround for using existing creative inventory item as icon for another group

dktapps committed:
Finalize dependency versions

dktapps committed:
CreativeInventoryEntry needs to clone the given item

dktapps committed:
Merge branch 'stable' into bedrock-1.21.60

dktapps committed:
Update Language dependency
